dali/kernels/reduce/reduce_gpu_impl.cuh

```diff
@@ -0,0 +1,1058 @@
+      const StageIn *tmp_in = wa.InputBuffer<StageIn>(stage.input_elements());
+      for (int i = 0; i < stage.num_samples(); i++)
+        ptrs[i] = tmp_in + stage.input_offsets[i];
+    } else {
+      assert((std::is_same<const StageIn, const In>::value));
+      for (int i = 0; i < stage.num_samples(); i++)
+        ptrs[i] = reinterpret_cast<const StageIn*>(ctx.input.data[i]);
+    }
+    return ptrs;
+  }
+
+  template <typename StageOut>
+  StageOut *const *OutputPtrs(Context &ctx, const ReductionStage &stage) const {
+    WorkArea &wa = ctx.work_area;
+    auto *ptrs = wa.ParamBuffer<StageOut*>(stage.num_samples());
+    if (!stage.is_last) {
+      StageOut *tmp_out = wa.OutputBuffer<StageOut>(stage.output_elements());
+      for (int i = 0; i < stage.num_samples(); i++)
+        ptrs[i] = tmp_out + stage.output_offsets[i];
+    } else {
+      assert((std::is_same<StageOut, Out>::value));
+      for (int i = 0; i < stage.num_samples(); i++)
+        ptrs[i] = reinterpret_cast<StageOut*>(ctx.output.data[i]);
+    }
+    return ptrs;
+  }
+
+  template <typename StageOut, typename StageIn>
+  auto PrepareSampleDescs(Context &ctx, const ReductionStage &stage) const {
+    using SampleDesc = ReduceSampleDesc<StageOut, StageIn>;
+    WorkArea &wa = ctx.work_area;
+    SampleDesc *samples = wa.ParamBuffer<SampleDesc>(stage.num_samples());
+
+    if (stage.index > 0) {
+      const StageIn *tmp_in = wa.InputBuffer<StageIn>(stage.input_elements());
+      for (int i = 0; i < stage.num_samples(); i++)
+        samples[i].in = tmp_in + stage.input_offsets[i];
+    } else {
+      assert((std::is_same<const StageIn, const In>::value));
+      for (int i = 0; i < stage.num_samples(); i++)
+        samples[i].in = reinterpret_cast<const StageIn*>(ctx.input.data[i]);
+    }
+
+    if (!stage.is_last) {
+      StageOut *tmp_out = wa.OutputBuffer<StageOut>(stage.output_elements());
+      for (int i = 0; i < stage.num_samples(); i++)
+        samples[i].out = tmp_out + stage.output_offsets[i];
+    } else {
+      assert((std::is_same<StageOut, Out>::value));
+      for (int i = 0; i < stage.num_samples(); i++)
+        samples[i].out = reinterpret_cast<StageOut*>(ctx.output.data[i]);
+    }
+
+    for (int i = 0; i < stage.num_samples(); i++) {
+      samples[i].n_inner = stage.shape[i].inner;
+      samples[i].n_outer = stage.shape[i].outer;
+      samples[i].n_reduced = stage.shape[i].reduced_in;
+      samples[i].num_macroblocks = stage.shape[i].reduced_out;
+      samples[i].macroblock_size = div_ceil(samples[i].n_reduced, samples[i].num_macroblocks);
+    }
+    return samples;
+  }
+
+  template <typename StageOut, typename StageIn, bool is_first, bool is_last>
+  void LaunchStage(Context &ctx, const ReductionStage &stage,
+                   ReductionKindTag<ReductionKind::All>) {
+    assert(!is_first || ctx.input.is_contiguous());
+    WorkArea &wa = ctx.work_area;
+
+    auto pre = This().GetPreprocessor(bool_const<is_first>());
```

Comment:
Can the `All` reduction be first?

Reply:
If there's just one sample in the input.